// verdaccio-demo-publish/src/menu.js
const states = {
  name: '', // 用户名称
  // 菜单路由
  dirs: [], // 目录级别数组，包含菜单级别数组
  permissions: [], // 按钮级别数组
  addRoutes: [], // 所拥有权限的异步路由+所有同步路由
  permissionRoutes: [] // 所拥有权限的异步路由+有权限的同步路由；也就是所拥有有权限的路由
}

/**
 * router路由格式
 * 不是菜单列表里的路由
 * {
 *    path: '/login',
 *    component: () => import('@/views/Login/index.vue'),
 *    hidden: true,
 *    name: 'login'
 * }
 * 
 * 一级菜单路由
 * {
 *    path: '/',
 *    component: Layout,
 *    children: [
 *      {
 *        path: 'home',
 *        component: () => import('@/views/home/index.vue'),
 *        name: 'home',
 *        meta: {
 *          title: '首页',
 *          icon: 'home',
 *          sid: 'home'
 *        }
 *       }
 *    ]
 * }
 * 
 * 二级菜单路由
 * {
 *    path: '/',
 *    redirect: '/third/menuOne',
 *    component: Layout,
 *    name: 'third',
 *    meta: {
 *      title: '功能菜单三',
 *      icon: 'third',
 *      sid: 'third'
 *    },
 *    children: [
 *      {
 *        path: 'menuOne',
 *        component: () => import('@/views/ThirdMenu/menuOne.vue'),
 *        name: 'menuOne',
 *        meta: {
 *          title: '功能菜单三-一',
 *          icon: 'menuOne',
 *          sid: 't-menuOne'
 *        }
 *      }
 *    ]
 * }
 * 
 */

/**
 * 用户权限roles结构
 * roles：[
 *    {}, // 用户角色一
 *    {
 *      id // 角色id
 *      type // 角色对象：admin 管理员 ordinary 普通
 *      permission: [ // 角色前端权限
 *        {}, // 权限一
 *        {
 *          type // 权限类别：1 菜单 2 按钮
 *          name // 权限名称
 *          sid  // 权限唯一标识，与router路由里的sid一致
 *          children: [ // 该权限下的子权限，一般按钮没有子权限，只有二级菜单会有子权限存在
 *            {}, // 子权限一
 *            {
 *              type // 权限类别：1 菜单 2 按钮
 *              name // 权限名称
 *              sid  // 权限唯一标识，与router路由里的sid一致
 *            }
 *          ]
 *        }
 *      ]
 *    }
 * ]
 * 
 */

/**
 * 
 * @param {*} dirs  // 菜单权限
 * @param {*} route // 路由信息
 * @returns Array
 */

const hasPermission = (dirs, route) => {
  let tmp = []
  route.children.forEach((child) => {
    if (child.meta && child.meta.title) {
      const idx = dirs.indexOf(child.meta.sid)
      // 如果子路由不包含在用户权限内则隐藏
      if (idx === -1) {
        child.hidden = true
      }
    } else {
      child.hidden = true
    }
  })
  tmp.push(route)
  return tmp
}

// routes:所有异步路由 menuList:登录用户拥有的菜单名称数组
/**
 * 
 * @param {*} routes    // asyncRoutes 所有异步路由
 * @param {*} menuList  // 用户所拥有的菜单权限
 * @returns Array
 */
const filterAsyncRoutes = (routes, menuList) => {
  let res = []
  routes.forEach((route) => {
    if (route.meta && route.meta.title && menuList.indexOf(route.meta.sid) > -1) {
      // 处理二级目录路由信息
      let tmp = hasPermission(menuList, route)
      // 将过滤后的路由信息添加到总的路由数组中
      res = res.concat(tmp)
    } else {
      // 无下拉菜单目录(子菜单)
      if (Array.isArray(route.children)) {
        let single = route.children[0]
        if (single.meta && single.meta.title && menuList.indexOf(single.meta.sid) > -1) {
          // single表示单独显示，而不是嵌套在下拉菜单中
          route.single = true
          res.push(route)
        }
      }
    }
  })
  return res
}

/**
 * 
 * @param {*} routes          // 所拥有权限的异步路由
 * @param {*} constantRoutes  // 所有同步路由
 * @param {*} menuList        // 菜单权限
 * @returns Array
 */
const getSetPermissionRoutes = (routes, constantRoutes, menuList) => {
  console.log('44444444444');
  let constantList = []
  // let permissionRoutes = []
  constantRoutes.forEach((item) => {
    if (item.children && item.children.length > 0) {
      let tmp = menuList.filter((val) => val === item.children[0].meta.sid)
      if (tmp.length > 0) {
        constantList.push(item)
      }
    }
    if (item.hidden) {
      constantList.unshift(item)
    }
  })
  states.permissionRoutes = constantList.concat(routes)
  return states.permissionRoutes
}

/**
 * 
 * @param {*} roles // 用户信息里的权限对象
 * @returns Object
 */
const formatMenuAndBtn = (roles) => {
  let MenuAndBtn = {
    menuArr: [],
    btnArr: []
  }
  roles.forEach((item) => {
    function transTree(val) {
      val.forEach((it) => {
        if (it.type === 1) {
          MenuAndBtn.menuArr.push(it.sid)
        }
        if (it.type === 2) {
          MenuAndBtn.btnArr.push(it.sid)
        }
        if (it.children) {
          transTree(it.children)
        }
        if (it.buttons) {
          transTree(it.buttons)
        }
      })
    }
    transTree(item.permission)
  })
  return MenuAndBtn
}

module.exports.hasPermission = hasPermission
module.exports.filterAsyncRoutes = filterAsyncRoutes
module.exports.getSetPermissionRoutes = getSetPermissionRoutes
module.exports.formatMenuAndBtn = formatMenuAndBtn
